### ISSUE TITLE
[BUG] inaccurate math functions in msys2-runtime 3.1.7

### ISSUE THREAD

**Issue by lhmouse** [msys2/MSYS2-packages#2210]
## Describe the bug

At least two inaccuracies have been identified in `msys2-runtime 3.1.7` which did not exist in `3.1.6`.

#### The `cos()` function

```c
#define _GNU_SOURCE
#include <stdio.h>
#include <math.h>

volatile double x = 6;
double fs, fc;

int main(void)
  {
    // `0.96017028665036597` or
    // `0.960170286650366` expected
    // note they are equivalent.
    sincos(x, &fs, &fc);
    printf("cos(%g)        = %.17g\n", x, cos(x));
    printf("via sincos(%g) = %.17g\n", x, fc);
  }
```

Compile and run with `gcc test.c -std=c99 && ./a.exe`.

The `cos()` function returns `0.9601702866503661` which is one ULP above the correct value. Note `sincos()` is not subject to this issue.

#### The `fma()` function

```c

#define _GNU_SOURCE
#include <stdio.h>
#include <math.h>

volatile double x =  0x1.0000000000003p-461;
volatile double y =  0x1.0000000000007p-461;
volatile double z = -0x1.000000000000Ap-922;

int main(void)
  {
    // `2.9204094392907018e-308` expected
    printf("fma = %.17g\n", fma(x, y, z));
  }
```

Compile and run with `gcc test.c -std=c99 && ./a.exe`.

The _only_ correct value is computed as follows:

```text
     0x                1.0000000000003 p-461
 x)  0x                1.0000000000007 p-461
----------------------------------------------
     0x                7 0000000000015
 +)  0x  1.0000000000003
----------------------------------------------
     0x  1.000000000000A 0000000000015 p-922
 -)  0x  1.000000000000A               p-922
----------------------------------------------
     0x  0.0000000000000 0000000000015 p-922 
   = 0x1.5p-1022
   ~ 2.9204094392907018e-308
```

But it now yields zero.

## Additional Context: Operating System, Screenshots

N/A

- OS: Windows 7 Professional SP1
- '...'

**Comment by lazka** [comment]
Where is this implemented, in newlib? Maybe @keith-packard knows as he worked on newlib math code recently.

**Comment by lhmouse** [comment]
Of course:  https://github.com/msys2/msys2-runtime/commits/msys2-3_1_7-release/newlib/libm

Is he active on GitHub?

**Comment by lhmouse** [comment]
I had some tests with `3.1.6` and identified that `fma()` did not produce the correct result even with an older version of runtime. I have [this test](https://github.com/lhmouse/asteria/blob/7e3d60f245a6ca8c1209578f553870a2148824ce/asteria/test/numeric.cpp#L246) which throws an exception if `fma()` produces an unexpected value, but for some unknown reason, with `msys2-runtime 3.1.6` the test program exited with zero, indicating success. The exception can be observed when running the test program in GDB. With `3.1.7` the test program exits with non-zero so the failure emerges.
